### ISSUE TITLE
[WARN ][logstash.codecs.line     ][main] Received an event that has a different character encoding than you configured.

### ISSUE THREAD

**Issue by myrsecurity** [pfelk/pfelk#112]
Hi, Im trying to workaround the message size limitation issue described in https://github.com/3ilson/pfelk/issues/111 by sending suricata logs via filebeat

So Im avoiding local Syslog registering for this exercise: Ive also configured another pfsense router externally 
router A inline mode filtering suricata -->Internet--> logstash input - filtering - output to elastic search. 
Filebeats input file:
```
filebeat.inputs:
- type: log
  paths:
    - "/var/log/suricata/*/eve.json*"
  fields_under_root: true
  fields:
    tags: ["suricata","json","pf","pfsense"]
- type: log
  paths:
    - "/var/syslog-ng/default.log"
  fields_under_root: true
  fields:
    tags: ["pfsense","pf"]


filebeat.config.modules:
  path: ${path.config}/modules.d/*.yml
  reload.enabled: false

output.logstash:
  hosts: ["destinationIPlogstash:5044"]

processors:
  - add_host_metadata: ~
  - add_cloud_metadata: ~


logging.metrics.enabled: false
logging.to_syslog: false
logging.to_files: true
logging.files:
  path: /var/log/filebeat
  name: filebeat.log
  keepfiles: 7
```
---------------

I have made a copy of the 01inputs to 02inputs and replaced the header to:
```
# 02-inputs.conf
input {

#    beats {
    port => 5044
#}

    tcp {
    port => 5044
  }

}

#codec => plain {
#charset => "UTF-8"  }


filter { 
#########################################################
# Adjust to match the IP address of pfSense or OPNsense #
#########################################################
  if [host] =~ /192\.168\.0\.1/ {
    mutate {
-------------------------------rest of the filter file. 


Communication between Router A reading Suricata logs and sending to Logstash Instance os working, but I receive in the logstash log

\\xD3}\\x87\\xF9\\xF9\\u0014\\xF6\\x92\\xA4\\a0\\u000F2\\e))\\u0016\\xAC\\xFFSU\\x8F\\xBE-jSr\\u000E\\xB0\\xAEx\\xBC\\\\\\xD3?a\\u0010\\x8DK\\xD8:\\u0019\\x91\\xA2\\xC5\\u001A\\xE3Y-\\u0016\\x84k\\u0006\\xCBŒ(\\xA2\\xB7\\xEA\\xE2M\\x88\\xE0\\xD3\\u0002mZch\\xC0\\xA7\\xEEE\\xE2\\xDD\\xC8\\xF5>\\xC9\\xC9k6p\\xE7KN^a\\x8D\\xA4\\xF7&2;\\xF9D\\xB5\\u0018", :expected_charset=>"UTF-8"}
[2020-04-25T14:43:08,601][WARN ][logstash.codecs.line     ][main] Received an event that has a different character encoding than you configured. {:text=>"!\\xC7\\\"\\xDE\\u0001X*\\xF8z\\u0015\\xDBd7\\a\\u0015$\\xE02\\xDE\\xD0t\\b>\\xEE\\xAC-'\\xC5\\xDC7\\x89L\\x92\\xDF\\xD4]\\xE9\\xBA\\u0016\\xF3V>\\x8D>z\\xEB܊w\\e\\xD8Orp\\x9E\\xB0o\\xCD \\xFD#\\xBD\\x81\\x9B\\xCF\\xC2\\u0003N\\xD9\\xE7\\xDD\\xFC*\\u001D\\u001C\\u0002[Ka\\xE1\\xFA\\x8D\\xAA:\\xF9J1\\xE5^\\x8F=\\xF8\\x8D)\\xB8\\x8A\\xCA\\u000E\\x8EF\\x85R\\xCE\\xEE\\fWp\\xE4yFY\\xEE'q\\x8B\\x9D\\u0016W3J\\xEB!k\\x9De\\xEEpX\\xD42g\\u0006\\xD6|\\u0011x\\x82\\xAA$\\xCA-Lo0\\x89ǧx봿\\u0019}\\xF6L\\u000Ey\\xBD#\\x8D\\x81-\\xC3sX\\xA0BQ\\\"\\xB8\\xE2\\x8A\\tlN\\e\\xDF\\\"\\xD1L\\x98T\\x90\\xB4\\u001D$\\x9A\\xA12\\u0004\\xD5X\\xA8\\u001A\\u0013mPe.\\xFB\\xB1f,\\u000E \\u0016v\\xE2R\\xBA4\\x84\\v\\xEC\\u000F\\xEE\\x92C\\x95\\x9A\\t\\xEAU9gv\\xA8\\u0003E\\xA2\\x92\\x86\\x8A\\xA4Rd\\xE38\\x9E,\\xF6b\\xEE\\xCBgٖxe5\\u0013|+\\xBA\\xC63\\a\\u0013\\xBCy\\xA6 !T\\xA5\\xC8\\xDF\\xC3", :expected_charset=>"UTF-8"}
[2020-04-25T14:43:08,601][WARN ][logstash.codecs.line     ][main] Received an event that has a different character encoding than you configured. {:text=>"\\xB8\\xB1i\\xE34\\u001E\\bH\\xE2\\xA2\\xD5ܘ", :expected_charset=>"UTF-8"}
[2020-04-25T14:43:08,601][WARN ][logstash.codecs.line     ][main] Received an event that has a different character encoding than you configured. {:text=>"\\xFAG", :expected_charset=>"UTF-8"}
[2020-04-25T14:43:19,211][WARN ][logstash.runner          ] SIGTERM received. Shutting down.
[2020-04-25T14:43:21,406][WARN ][logstash.codecs.line     ][main] Received an event that has a different character encoding than you configured. {:text=>"X͘\\xAA\\x8B,\\x80>f5\\aXeXg;\\x84l\\t\\u0012\\xF7\\xCE-&\\x89\\xBC\\xF1\\xD9\\xDFQ\\u0015\\r\\x8B\\xEFJ\\xB71\\x9B\\x8D\\r\\xB3\\xE2uF\\x8A\\xD3-l\\x8B\\xEFX\\xE9>\\x86רj\\x8È\\x9D\\xB7\\xB0ISu\\x8Cs.\\xF42B\\xBB\\f\\t\\x88W\\x8A\\x94FFi\\xF0\\xE7#]rP\\xA9b\\xC9\\xE4\\xCB\\xFF\\e\\u0000\\u0000\\xFF\\xFF\\u0017a\\xAF\\xC2", :expected_charset=>"UTF-8"}
[2020-04-25T14:43:23,653][INFO ][logstash.javapipeline    ] Pipeline terminated {"pipeline.id"=>"main"}
[2020-04-25T14:43:23,696][INFO ][logstash.runner          ] Logstash shut down.
```
Essentially the expected charset in Logstash is UTF-8 while the one sending 

```
xBCy\\xA6 !T\\xA5\\xC8\\xDF\\xC3", :expected_charset=>"UTF-8"}
[2020-04-25T14:43:08,601][WARN ][logstash.codecs.line     ][main] Received an event that has a different character encoding than you configured. {:text=>"\\xB8\\xB1i\\xE34\\u001E\\bH\\xE2\\xA2\\xD5ܘ", :expected_charset=>"UTF-8"}
```

Ive tried few settings on the way but didnt work
```
#codec => plain {
#charset => "UTF-8"  }
```

Thank you guys

**Comment by revere521** [comment]
Thanks for doing that while i was asking you to do that , but before i posted it lol

please put your pasted code inside code headers (select the text and click the <> icon in the edit window)  so its easier to read 

you may need to try codec => json  and charset => utf-8 ?

**Comment by revere521** [comment]
you will probably need to do this so its all together in one section before the rest of the inputs: 
```

beats {
    port => 5044
         codec => json {
         charset => "utf-8" }
}
       
```

**Comment by myrsecurity** [comment]
with that input configuration I dont receive any data:

Im rebooting the source firewall in case Im hitting a bug I read elsewhere

**Comment by a3ilson** [comment]
@myrsecurity - you can simply run a tcpdump command on the node receiving the logs.  This is a simple sanity check to see if pfSense is sending vs rebooting

**Comment by myrsecurity** [comment]
I suspect I need to amend the firewall IP configuration in inputs and set the external IP address rather than the internal

**Comment by myrsecurity** [comment]
I see lots of DNS UDP 53 traffic - 

And also I see traffic on 5044


16:36:02.702028 IP logs.pikoleto.com.5044 > m85-94-184-58.andorpac.ad.40575: Flags [.], ack 49081, win 896, options [nop,nop,TS val 1566739434 ecr 3480772789], length 0
16:36:02.708040 IP logs.pikoleto.com.5044 > m85-94-184-58.andorpac.ad.40575: Flags [P.], seq 73:79, ack 49081, win 896, options [nop,nop,TS val 1566739440 ecr 3480772789], length 6
16:36:02.708052 IP m85-94-184-58.andorpac.ad.40575 > logs.pikoleto.com.5044: Flags [.], ack 79, win 513, options [nop,nop,TS val 3480772824 ecr 1566739440], length 0
16:36:03.676372 IP m85-94-184-58.andorpac.ad.40575 > logs.pikoleto.com.5044: Flags [P.], seq 49081:50292, ack 79, win 513, options [nop,nop,TS val 3480773792 ecr 1566739440], length 1211
16:36:03.706658 IP logs.pikoleto.com.5044 > m85-94-184-58.andorpac.ad.40575: Flags [P.], seq 79:85, ack 50292, win 896, options [nop,nop,TS val 1566740438 ecr 3480773792], length 6
16:36:03.706681 IP m85-94-184-58.andorpac.ad.40575 > logs.pikoleto.com.5044: Flags [.], ack 85, win 513, options [nop,nop,TS val 3480773822 ecr 1566740438], length 0

**Comment by myrsecurity** [comment]
IVe removed the IP address by the way - and replaced with pikoleto

**Comment by a3ilson** [comment]
You can omit the IP address in the 01-inputs.conf file.  However, the IP address should be the IP address (pfSense) that is sending the logged traffic (typically this is the internal IP).

**Comment by myrsecurity** [comment]
:response=>{"index"=>{"_index"=>"pf-2020.04.25", "_type"=>"_doc", "_id"=>"nzbZsXEBmryOZlzwlZgQ", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [source] tried to parse field [source] as object, but found a concrete value"}}}}
[2020-04-25T16:59:18,935][WARN ][logstash.outputs.elasticsearch][main] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"pf-2020.04.25", :routing=>nil, :_type=>"_doc"}, #<LogStash::Event:0x628e7dfb>], :response=>{"index"=>{"_index"=>"pf-2020.04.25", "_type"=>"_doc", "_id"=>"oDbZsXEBmryOZlzwlZgT", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [source] tried to parse field [source] as object, but found a concrete value"}}}}

**Comment by myrsecurity** [comment]
so now it seems the logs are received, processed but failed to index it

**Comment by myrsecurity** [comment]
< ion=>["index", {:_id=>nil, :_index=>"pf-2020.04.25", :routing=>nil, :_type=>"_doc"}, #<LogStash::Event:0xf8cfb7f>], :response=>{"index"=>{"_index"=>"pf-2020.04.25", "_type"=>"_doc", "_id"=>"7iHwsXEBEMM_zmy1eH0m", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [source] tried to parse field [source] as object, but found a concrete value"}}}}
[2020-04-25T17:24:18,874][WARN ][logstash.outputs.elasticsearch][main] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"pf-2020.04.25", :routing=>nil, :_type=>"_doc"}, #<LogStash::Event:0x37b53d63>], :response=>{"index"=>{"_index"=>"pf-2020.04.25", "_type"=>"_doc", "_id"=>"7yHwsXEBEMM_zmy1eH0m", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [source] tried to parse field [source] as object, but found a concrete value"}}}}
 >

**Comment by a3ilson** [comment]
@myrsecurity - Can you provide the raw output?  I am going to assume that beats didn't filter as pfELK expected/intended.

**Comment by myrsecurity** [comment]
Ive now removed the extra tags in the beat input side:
filebeat.inputs:
- type: log
  paths:
    - "/var/log/suricata/*/eve.json*"
  fields_under_root: true
  fields:
    tags: ["pf"]
- type: log
  paths:
    - "/var/syslog-ng/default.log"
  fields_under_root: true
  fields:
    tags: ["pf"]

leaving only pf ( from a mixture of pfsense, pf, suricata, etc ) 

restarted services across the board, but got the same output in logstash 
5][WARN ][logstash.outputs.elasticsearch][main] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"pf-2020.04.25", :routing=>nil, :_type=>"_doc"}, #<LogStash::Event:0x45c17b0b>], :response=>{"index"=>{"_index"=>"pf-2020.04.25", "_type"=>"_doc", "_id"=>"oBn6sXEBpnFObJACAMQ2", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [packet] tried to parse field [packet] as object, but found a concrete value"}}}}
[2020-04-25T17:34:43,536][WARN ][logstash.outputs.elasticsearch][main] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"pf-2020.04.25", :routing=>nil, :_type=>"_doc"}, #<LogStash::Event:0x6d869a86>], :response=>{"index"=>{"_index"=>"pf-2020.04.25", "_type"=>"_doc", "_id"=>"oRn6sXEBpnFObJACAMQ2", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [packet] tried to parse field [packet] as object, but found a concrete value"}}}}
[2020-04-25T17:34:43,536][WARN ][logstash.outputs.elasticsearch][main] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"pf-2020.04.25", :routing=>nil, :_type=>"_doc"}, #<LogStash::Event:0x4b6ae36c>], :response=>{"index"=>{"_index"=>"pf-2020.04.25", "_type"=>"_doc", "_id"=>"ohn6sXEBpnFObJACAMQ2", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [packet] tried to parse field [packet] as object, but found a concrete value"}}}}
[2020-04-25T17:34:43,536][WARN ][logstash.outputs.elasticsearch][main] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"pf-2020.04.25", :routing=>nil, :_type=>"_doc"}, #<LogStash::Event:0x7865ea05>], :response=>{"index"=>{"_index"=>"pf-2020.04.25", "_type"=>"_doc", "_id"=>"oxn6sXEBpnFObJACAMQ2", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [packet] tried to parse field [packet] as object, but found a concrete value"}}}}
[2020-04-25T17:34:44,091][WARN ][logstash.outputs.elasticsearch][main] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"pf-2020.04.25", :routing=>nil, :_type=>"_doc"}, #<LogStash::Event:0x5a13f424>], :response=>{"index"=>{"_index"=>"pf-2020.04.25", "_type"=>"_doc", "_id"=>"pRn6sXEBpnFObJACAsR3", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [packet] tried to parse field [packet] as object, but found a concrete value"}}}}
[2020-04-25T17:34:44,091][WARN ][logstash.outputs.elasticsearch][main] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"pf-2020.04.25", :routing=>nil, :_type=>"_doc"}, #<LogStash::Event:0x4ba4ead4>], :response=>{"index"=>{"_index"=>"pf-2020.04.25", "_type"=>"_doc", "_id"=>"phn6sXEBpnFObJACAsR3", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [packet] tried t

**Comment by myrsecurity** [comment]
How can I give you a raw output?

**Comment by myrsecurity** [comment]
Ive forgot to mention , in one of my previous attempts I managed to parse this to Elastic
![image](https://user-images.githubusercontent.com/53452334/80284573-b7773400-871f-11ea-9eea-7da874872187.png)

**Comment by myrsecurity** [comment]
You can see that the message is completely unreadable, I got a warning in Logstash referring to charset.

**Comment by a3ilson** [comment]
@myrsecurity - Have you read over #75?  There is info and pipeline config files for beats and what your trying to accomplish.

**Comment by myrsecurity** [comment]
Thank you - Yes I have read it but not implemented yet. It seems there is a slight important difference: my current log output by filebeats is traveling over the internet (unencrypted by now, once I get this working Ill add the SSL layer ) until they reach logstash. 
On 75 I understand everything is on the local network. 

Will give it another go, Im sure we arent far is just a matter of understanding the right output format beats is doing

**Comment by myrsecurity** [comment]
Im also thinking on a peer to peer tunnel between pfsense and ELK so I can output on syslog

**Comment by myrsecurity** [comment]
thanks for your help, I never expect this level of support btw

**Comment by a3ilson** [comment]
No worries, wish we could be of more assistance.  Once you get if figured out and in addition to #75, I'll add as an alternative configuration via Wiki Page....so if you don't mind contributing- take good notes. 

In the meantime feel free to keep posting and we'll help/assist as able.

**Comment by myrsecurity** [comment]
thank you

**Comment by myrsecurity** [comment]
There is something similar here btw https://villekaaria.eu/2019/03/24/suricata-logs-to-logstash-with-filebeat-on-pfsense-2-4/ 

I used this resource before, yes you can put the suricata logs in ELK, but the whole dashboard bit where pfelk has an advantage is missing.

**Comment by a3ilson** [comment]
@myrsecurity - were you able to make any headway?

**Comment by myrsecurity** [comment]
nope, using syslog to remote logstash over vpn : filebeats module for FreeBSD works well to pfsense->elastic. The problem is that you need to recreate all of the dashboards and reports again as elastic it doesn't pick up out of the box suricata traffic and puts it on 'suricata events dashboard' 
cheers

**Comment by a3ilson** [comment]
@myrsecurity - I am linking this with issue #121.  The remote syslog is working and sufficient.  The issue will be snort/suricate and other log data being sent through syslog. This (beats on pfSense) is one method to circumvent the limitation of the enforced RFCs but I'm hoping to come across or develop something else that may be able to provide an alternative/easier workaround.

**Comment by a3ilson** [comment]
@myrsecurity - this issue was solved by @TehMuffinMoo within issue #111.  The solution was also posted within the wiki https://github.com/3ilson/pfelk/wiki/How-To:-Suricata-on-pfSense
